crates/subspace-farmer/src/cluster/nats_client.rs

```diff
@@ -0,0 +1,433 @@
+    /// request
+    const SUBJECT: &'static str;
+}
+
+/// Generic broadcast message.
+///
+/// Broadcast messages are sent by an instance to (potentially) an instance-specific subject that
+/// any other app can subscribe to. The same broadcast message can also originate from multiple
+/// places and be de-duplicated using [`Self::deterministic_message_id`].
+pub trait GenericBroadcast: Encode + Decode + fmt::Debug + Send + Sync + 'static {
+    /// Broadcast subject with optional `*` in place of application instance sending broadcast
+    const SUBJECT: &'static str;
+
+    /// Deterministic message ID that is used for de-duplicating messages broadcast by different
+    /// instances
+    fn deterministic_message_id(&self) -> Option<HeaderValue> {
+        None
+    }
+}
+
+/// Subscriber wrapper that decodes messages automatically and skips messages that can't be decoded
+#[derive(Debug, Deref, DerefMut)]
+#[pin_project::pin_project]
+pub struct SubscriberWrapper<Message> {
+    #[pin]
+    #[deref]
+    #[deref_mut]
+    subscriber: Subscriber,
+    _phantom: PhantomData<Message>,
+}
+
+impl<Message> Stream for SubscriberWrapper<Message>
+where
+    Message: Decode,
+{
+    type Item = Message;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        match self.project().subscriber.poll_next_unpin(cx) {
+            Poll::Ready(Some(message)) => match Message::decode(&mut message.payload.as_ref()) {
+                Ok(message) => Poll::Ready(Some(message)),
+                Err(error) => {
+                    warn!(
+                        %error,
+                        message_type = %type_name::<Message>(),
+                        message = %hex::encode(message.payload),
+                        "Failed to decode stream message"
+                    );
+
+                    Poll::Pending
+                }
+            },
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+/// NATS client wrapper that can be used to interact with other Subspace-specific clients
+#[derive(Debug, Clone, Deref)]
+pub struct NatsClient {
+    client: Client,
+}
+
+impl From<Client> for NatsClient {
+    fn from(client: Client) -> Self {
+        Self { client }
+    }
+}
+
+impl NatsClient {
+    /// Create new instance by connecting to specified addresses
+    pub async fn new<A: ToServerAddrs>(addrs: A) -> Result<Self, async_nats::Error> {
+        Ok(Self {
+            client: async_nats::connect(addrs).await?,
+        })
+    }
+
+    /// Make request and wait for response
+    pub async fn request<Request>(
+        &self,
+        request: &Request,
+        instance: Option<&str>,
+    ) -> Result<Request::Response, RequestError>
+    where
+        Request: GenericRequest,
+    {
+        let subject = if let Some(instance) = instance {
```

Comment:
This code and all other subject creations will benefit from the helper function. It will remove a dozen copied code snippets.

Reply:
There are just 3 places with this in the module :slightly_smiling_face:

Reply:
5 actually - there are several ways to deal with it: either a common function for requests and notifications or a separate for both but the latter will have only an instance as a parameter.